uint8 FAIL_FAST=0
uint8 CONTINUE=1
uint8 mode